### PR TITLE
Simplify `ServicePluginManager`

### LINKED ISSUE #3979: Check requirements for non-service plugins
In order to check requirements for non-service plugins (e.g. content) at the same time as checking requirements for service plugins, we must be able to extract these requirements from `Data` (with `ProjectConfiguration` being an example root).

## To do
- Move `ServicePluginDefinition.requires` to `PluginDefinition.requires`
- Add `requires` to `ServicePluginProvider.__init__()`. This should replace `Project.__init__()`'s `extensions`
- Add an API for `Data` classes to expose requirements and a reusable implementation to to this for configured plugins.

## Blocked by 
- https://github.com/bartfeenstra/betty/pull/3978

### PR DESCRIPTION
**Pull request by bartfeenstra**
This fixes https://github.com/bartfeenstra/betty/issues/3979

### CODE REVIEW

File: betty/app/__init__.py [hunk 1]
```diff
@@ -46,12 +46,15 @@
     from betty.plugin import PluginDefinition
     from betty.plugin.discovery import ResolvableDiscovery
     from betty.user import User
+type AppServicePlugin = AssetDefinition | RateLimitDefinition
```
Comment (bartfeenstra) [inline betty/app/__init__.py:51]: Two newlines before

File: betty/service/plugin.py [hunk 2]
```diff
@@ -229,27 +229,51 @@ def requires(self) -> Iterable[Requirement]:
...
+]
+type SupportPlugins = Iterable[
+    PluginManufacturer[PluginDefinition, Plugin[PluginDefinition]]
+    | type[Plugin[PluginDefinition]]
+]
+
+
 @final
-class ServicePluginManager(ManagedLifeCycle):
+class ServicePluginManager[ServicePluginTypesT: ServicePluginDefinition](
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: Make this class private. Move its tests to `TestServicePluginProvider`

File: betty/service/plugin.py [hunk 3]
```diff
@@ -229,27 +229,51 @@ def requires(self) -> Iterable[Requirement]:
...
     """
 
     def __init__(
         self,
-        service_plugin_manufacturers: ServicePluginManufacturers | None,
-        *,
         services: ServiceLevel,
+        plugin_types: Collection[type[ServicePluginTypesT]],
+        plugins: ServicePlugins[ServicePluginTypesT] = (),
+        support: SupportPlugins = (),
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: Include the support plugins' requirements in `_expand_requires()`

File: betty/service/plugin.py [hunk 4]
```diff
@@ -368,14 +392,37 @@ def __iter__(self) -> Iterator[type[ServicePluginDefinition]]:
...
+    def __init__(
+        self,
+        *args: Any,
+        service_plugin_services: ServiceLevel,
+        service_plugin_types: Collection[type[ServicePluginTypesT]] = (),
+        service_plugins: ServicePlugins[ServicePluginTypesT] = (),
+        **kwargs: Any,
+    ):
+        super().__init__(*args, **kwargs)
+        self._service_plugin_types = service_plugin_types
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: Prefix with double underscores

File: betty/service/plugin.py [hunk 5]
```diff
@@ -229,27 +229,51 @@ def requires(self) -> Iterable[Requirement]:
 ]
 
 
+type ServicePlugins[
+    ServicePluginTypesT: ServicePluginDefinition = ServicePluginDefinition
+] = Iterable[
+    PluginManufacturer[ServicePluginTypesT, Plugin[ServicePluginTypesT]]
+    | type[Plugin[ServicePluginTypesT]]
+]
+type SupportPlugins = Iterable[
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: Use a common private base type to ensure both are in sync

File: betty/service/plugin.py [hunk 6]
```diff
@@ -229,27 +229,51 @@ def requires(self) -> Iterable[Requirement]:
...
     """
     The service plugin manager.
     """
 
     def __init__(
         self,
-        service_plugin_manufacturers: ServicePluginManufacturers | None,
-        *,
         services: ServiceLevel,
+        plugin_types: Collection[type[ServicePluginTypesT]],
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: `service_plugin_types`, `service_plugins`, and `supported_plugins`

File: betty/service/plugin.py [hunk 7]
```diff
@@ -221,35 +221,52 @@ def requires(self) -> Iterable[Requirement]:
...
 ]
+type ServicePlugins[
+    ServicePluginT: ServicePluginDefinition = ServicePluginDefinition
+] = _Plugins[ServicePluginT]
+type SupportPlugins = _Plugins[PluginDefinition]
 
 
 @final
-class ServicePluginManager(ManagedLifeCycle):
+class ServicePluginManager[ServicePluginTypesT: ServicePluginDefinition](
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: I don't think this needs to be generic after all.

File: betty/service/plugin.py [hunk 8]
```diff
@@ -368,14 +385,37 @@ def __iter__(self) -> Iterator[type[ServicePluginDefinition]]:
         return iter(self._service_plugins)
 
 
-class ServicePluginProvider(ManagedLifeCycle, ABC):
+class ServicePluginProvider[ServicePluginTypesT: ServicePluginDefinition](
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: I don't think this needs to be generic after all.

File: betty/service/plugin.py [hunk 9]
```diff
@@ -368,14 +385,37 @@ def __iter__(self) -> Iterator[type[ServicePluginDefinition]]:
...
+    ManagedLifeCycle
+):
     """
     A service plugin provider.
     """
 
+    def __init__(
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: This must be generic over the plugin definition type instead.

File: betty/app/__init__.py
Comment (bartfeenstra) [inline betty/app/__init__.py:70]: Here and for Project, add a `service_plugins` kwarg.

File: betty/service/plugin.py [hunk 10]
```diff
@@ -221,35 +221,52 @@ def requires(self) -> Iterable[Requirement]:
...
     The service plugin manager.
     """
 
     def __init__(
         self,
-        service_plugin_manufacturers: ServicePluginManufacturers | None,
-        *,
         services: ServiceLevel,
+        plugin_types: ServicePluginTypes[ServicePluginTypesT],
+        plugins: ServicePlugins[ServicePluginTypesT] = (),
```
Comment (bartfeenstra) [inline betty/service/plugin.py]: Add test coverage to ensure that an error is raised if duplicate plugins are given.
